### PR TITLE
Interactive Map

### PR DESCRIPTION
**Pull request by rapotek**
The repository seems to be almost stale, so...

This is a proposition of changes to an existing Dynamic Map, together with corresponding changes is My Neighbourhood. Because of the nature of changes, as perceived from an external user point, I named it an Interactive Map.

IMPORTANT: Because there is a bunch of changes which I consider should be validated by someone, and not only developers but OC Team and advanced users too, the whole set of changes is made to coexist with an official My Neighbourhood, but replacing it after validation should require few changes only, apart from files relocation. To assure this coexistence, some MyNeighbourhood files have been simply copied to another location, while another ones has been only slightly changed.

What is new from a user point of view:

- A popup with marker details, showed by clicking on map marker, allows now to see not only a top-most marker, but also "move" to the next/previous marker being under the exact coordinates of the "clicked one",
- The visibility and order of markers shown on map is consistent with currently set order and visibility of My Neighbourhood "sections" (latest caches, latest logs etc.) and changing the section order or visibility adjusts the map appearance on client side. Ordering: the most top section markers are displayed on top, the second one are next etc.
- The visibility and order changes should apply to currently opened popup contents too.
- If the marker belongs to a named section, the section title is shown in the top part of popup box.
- When the mouse/cursor is moved over a row containing a particular entry, a corresponding marker is "highlighted" on the map and the "highlight" is toned down on mouse/cursor exit.

What is new from a developer point of view (main things):

- All markers within MapModel are now placed within sections. When a marker does not belong to any section, it should be placed within a special, `_DEFAULT_` section.
- Each section has a corresponding layer created in map and the visibility and ordering are layers' operations.
- `DynamicMap` itself (now `InteractiveMap`), its tools etc. have been rewritten to the more objective version, with prototyping where there is a chance for more than one instance creation.
- `InteractiveMapServices` global object should be now an entry point for every external usage.
- Markers have been rewritten to an objective, prototyped version, with inheritance introduced, and grouped within named 'marker family', where each defined family should have its own styling. Only a 'simple' family, using icons "computed" on server side is available now, but an implementation of more advanced 'okapi' icons is in progress.
- A more detailed information is passed with each marker within `markerData`, it is a step to introduce client-side 'okapi' icons too.

Above changes should be available for validation for OCTeam, Advanced Users and System Admins under '/MyNbhInteractive' relative url.

Should work (tested) on Firefox 78, Chrome 83, Edge 44. Does not show latest logs on IE 11 because of SVG icons, but current My Neighbourhood does not show them too.

### CODE REVIEW

**Comment by kojoty** [comment]
hi Rapotek, nice to see your PR :)

on the first sight it seems your changes are mostly in context of UI so we can run this code paralel to production site - right? 
We can run "test" version of the site running on production DB but with code base from this PR only for some numbers of testers?

It could allow all of us to better understand your changes.

I have very limited time resources for OCPL last months so please expect some delays from time to time - but i still try to support OCPL development :)

**Comment by rapotek** [comment]
> on the first sight it seems your changes are mostly in context of UI so we can run this code paralel to production site - right?
> We can run "test" version of the site running on production DB but with code base from this PR only for some numbers of testers?

Yep, as I wrote above in other words:

- The map client-side code is fully independent from existing dynamic map, so they can coexist together in the same environment without interfering with each other. In fact while testing I had both old and new My Neighbourhood opened in separate browser tabs in the same time and they both worked.
- The changes in common, existing files are only necessary additions used in this PR code and nowhere else. They do not change existing code behaviour in any way.
- The new My Neighbourhood is available currently only for OCTeam, Advanced Users and System Admins, other users are redirected to the main page.

In summary, from my point of view the changes can be incorporated into a production site without disrupting anything for common users. However I would wait with it until the weekend passes, as usual, to be compliant with our guidelines.

File: src/Models/GeoCache/GeoCacheCommons.php [hunk 1]
```diff
@@ -467,4 +469,26 @@ public static function GetCacheUrlByWp($ocWaypoint)
...
         return '/viewcache.php?wp=' . $ocWaypoint;
     }
+
+    /**
+     * Returns a JSON structure containing a (constant_name, constant_value)
+     * pairs for all defined geocache statuses
+     *
+     * @return string JSON structure, ready for use in javascript f.ex.
+     */
+    public static function CacheStatusListJson()
```
Comment (kojoty) [inline src/Models/GeoCache/GeoCacheCommons.php:479]: jeśli dobrze rozumiem ideę to moze prosciej jest zrobić 
 json_encode(geocacheCommons::CacheStatusArray())
??

File: src/Models/GeoCache/GeoCacheLogCommons.php [hunk 2]
```diff
@@ -235,4 +237,26 @@ public static function getLogTypeTplKeys(
...
         return $result;
     }
+
+    /**
+     * Returns a JSON structure containing a (constant_name, constant_value)
+     * pairs for all defined log types
+     *
+     * @return string JSON structure, ready for use in javascript f.ex.
+     */
+    public static function LogTypeListJson()
```
Comment (kojoty) [inline src/Models/GeoCache/GeoCacheLogCommons.php:247]: tutaj podobnie...

File: src/Models/GeoCache/GeoCacheCommons.php (hunk 1 above)
Comment (rapotek) [reply src/Models/GeoCache/GeoCacheCommons.php:479]: Wynik nie jest ten sam.
`json_encode(GeoCacheCommons::CacheStatusArray())` daje w wyniku: `[1,2,3,4,5,6]`,
tymczasem ta metoda ma dawać: `{"STATUS_READY":"1","STATUS_UNAVAILABLE":"2","STATUS_ARCHIVED":"3","STATUS_WAITAPPROVERS":"4","STATUS_NOTYETAVAILABLE":"5","STATUS_BLOCKED":"6"}`.
Podobnie w `GeoCacheLogCommons`

**Comment by rapotek** [comment]
DO NOT MERGE NOW!

I worked on okapi icons and discovered that the determination of most top icon displayed is wrong. It needs deeper refactoring inside `interactiveMap.js`, I am working on it but it will take time, so please leave this PR until further notice from me.

**Comment by rapotek** [comment]
Sorry for no updates for so long, I had problems with my dev environment accessibility.

I decided to close this PR and open a new one with issues mentioned above resolved, because in the mean time I managed to finish OKAPI-based style markers. The new PR will be soon (maybe tomorrow) and it will contain fixes here as well as a bunch of files for a new style, OKAPI-based. A user will be able to select a more preferred style personally.
